### ISSUE TITLE
Extending the File > New dialogue even more

### ISSUE THREAD

**Issue by rdstern** [africanmathsinitiative/R-Instat#4439]
I now propose 3 of our radio buttons for the File > New dialogue.

One is just **New** and is the current dialogue. The second could be **Wakefield** and is discussed in another issue, see #4347.  The third is called **Data** and is a facility to add lines of data automatically see #4438. 
 The third option should be an easy addition?  It needs to have the number of rows (as does the first option), but not the number of columns.  It also needs the name of the resulting data frame.  There also is a wide text field for multiple lines, into which you can paste R commands.
There could also be a second, smaller field - perhaps on all options - where you could include a data frame label.

Then the R code needed is as follows:

**data2 <- data.frame(data=matrix(data=NA, nrow=20, ncol=0))
data2<-within( data2,{**
block<-c(rep("I",7),rep("II",6),rep("III",7))
trt<-c("A","B","C","D","g","k","l","A","B","C","D","e","i","A","B","C","D","f","h","j")
yield<-c(83,77,78,78,70,75,74,79,81,81,91,79,78,92,79,87,81,89,96,82)**}**
**)**

Here the data2 is the data frame name.  And 20 is the number of rows.  The three lines pasted into the text field are those starting block, trt, yield, see #4438.  

Piece of cake?
What isn't so clear is what messages will be given if those lines contain a mistake.  This can be checked partly by leaving out say one value from the yield column, so it is only of length 19 - in this example.

**Comment by rdstern** [comment]
In the proposed new guide on giving commands I have used the example in agricolae for BIB.  This has a data frame of 30 observations with these being the lines:

The first line - with a comment in front, could usefully be copied into the label field for the data frame.

# Example Design of Experiments. Robert O. Kuehl. 2nd. Edition. 2001 
run<- gl(10,3) 
psi<- c(250,325,475,250,475,550,325,400,550,400,475,550,325,475,550,250,400,475,250,325,400,250,400,550,250,325,550,325,400,475) 
monovinyl<- c(16,18,32,19,46,45,26,39,61,21,35,55,19,47,48,20,33,31,13,13,34,21, 30,52,24,10,50,24,31,37) 

In the script window, these commands run fine, but only if each is on a single line.  I wonder if that condition can be relaxed here?  Not if it takes too much time.

**Comment by dannyparsons** [comment]
Running one line at a time is the only way we can do the script window sensibly really so this wouldn't be ready to modify.

**Comment by dannyparsons** [comment]
Also, I wouldn't recommend doing this with the script window . This is because the vectors or date frame you create live outside the Instat object. You would have to know the instat object command to get this added like another data frame. Without this the data frame can't appear in the spreadsheet and also isn't accessible in dialogs. I would guess this might be confusing to have in a guide.

**Comment by rdstern** [comment]
I am not assuming using the script window.  I am suggesting that the **File > New** dialogue has the possibility that you could type (or more likely paste) commands into its own editing window and that they generate data.  

It would have one line per variable

When you press Ok it would send the set of lines, possibly with an add between each line.  



Here is an example using the data I have currently for this command guide.  I used the current File > New, then the Prepare > Column: Generate: Enter dialogues to construct this.  The idea is that this would be the code run through the new option in the File > New dialogue, which becomes another of our dangerous "halfway" dialogues:

#### Code generated by the dialog, New Data Frame

data1 <- data.frame(data=matrix(data=NA, nrow=30, ncol=0))
InstatDataObject$import_data(data_tables=list(data1=data1))

rm(data1)
#### Code generated currently by the dialog, Enter with each command on one line!

**run<-gl(10,3)**
InstatDataObject$add_columns_to_data(col_data=run, require_correct_length=FALSE, data_name="data1", col_name="run")
rm(run)
**psi<-c(250,325,475,250,475,550,325,400,550,400,475,550,325,475,550,250,400,475,250,325,400,250,400,550,250,325,550,325,400,475)**
InstatDataObject$add_columns_to_data(col_data=psi, require_correct_length=FALSE, data_name="data1", col_name="psi")
rm(psi)
**monovinyl<-c(16,18,32,19,46,45,26,39,61,21,35,55,19,47,48,20,33,31,13,13,34,21,30,52,24,10,50,24,31,37)**
InstatDataObject$add_columns_to_data(col_data=monovinyl, require_correct_length=FALSE, data_name="data1", col_name="monovinyl")
rm(monovinyl)

In the code above, just the 3 data lines are typed into the File > New editing window.  There wouldn't be many checks I assume - but perhaps there could be some.  For example that in the **File > New** editing window each line would contain the string "<-"

**Comment by dannyparsons** [comment]
Then I'm happy with that. File > New is not very useful at the moment so having these options which are not straightforward to use but more useful is better than what we have.

These expressions go straight into the data frame command, this wouldn't be done in the same way as adding columns.

```
data1 <- data.frame(run = gl(10,3), 
                    psi = c(250,325,475,250,475,550,325,400,550,400,475,550,325,475,550,250,400,475,250,325,400,250,400,550,250,325,550,325,400,475), 
                    monovinyl = c(16,18,32,19,46,45,26,39,61,21,35,55,19,47,48,20,33,31,13,13,34,21,30,52,24,10,50,24,31,37)
                    )
```

We can then gradually make it easier to build up the commands, e.g. with keyboards. Already I would have a table structure for this so you enter the name of the column and then the expression for the values. 

| Name      | Values |
|-----------|----------------------------------------------------------------------------------------------------------------------------|
| run       | gl(10,3)                                                                                                                   |
| psi       | c(250,325,475,250,475,550,325,400,550,400,475,550,325,475,550,250,400,475,250,325,400,250,400,550,250,325,550,325,400,475) |
| monovinyl | c(16,18,32,19,46,45,26,39,61,21,35,55,19,47,48,20,33,31,13,13,34,21,30,52,24,10,50,24,31,37)                               |

**Comment by rdstern** [comment]
That looks much better.  As you say it also turns **File > New** into something more useful.

Of course the next step from your table could be to add a keyboard as we have in the **Prepare > Column: Generate > Enter dialogue**.
But even with the table it is a good step towards becoming a "proper" dialogue and not simply a simplistic "halfway" one.
It works for one important use I envision, namely to input examples from R packages that give the data in this way.

**Comment by dannyparsons** [comment]
If you want to be able to copy example from elsewhere then there could be the table option (default) and a command option where you get a large text box to enter any expression which evaluates as a data frame. The default text could be `data.frame()`.

**Comment by Patowhiz** [comment]
Does this dialog exist? I seem not to find it.

**Comment by rdstern** [comment]
It certainly exists - though not sure of the name.  In the current menu **File > New** just gives an empty data frame of the specified dimensions.  It is the first dialogue in the File Menu.

**Comment by dannyparsons** [comment]
Some thought is needed in how the table part will work. I'm not even sure which controls to use. I see it having text underneath saying add which adds another row to the table to define another column. So it needs to be expandable. I would suggest it's contained in something so that maybe 5 rows are visible and any more are seen by scrolling.

I'm pretty clear on the overall layout of the dialog but maybe Patrick you should send a quick sketch before starting on this to check we're all on the same page

**Comment by Patowhiz** [comment]
Yep. I'll have a sketch by tomorrow morning

**Comment by rdstern** [comment]
This is exciting.  It will probably do more than I expected on helping to teach users R, while also continuing with their use of R-Instat.

I am still interested in one use being to be able to access the examples from the other packages, while also being able to see the data in a data frame in R-Instat.

I wonder if we should include the keyboard from the **Prepare > Column: Generate > Enter** dialogue in the design from the start? It could be improved, but it could also be used to add extra columns - it is another "halfway" dialogue, so I suggest it could be the identical keyboard in both dialogues.

**Comment by rdstern** [comment]
Here is a second example of data - from **plotrix**.
**fake some reference data**
ref<-rnorm(30,sd=2)
**add a little noise**
model1<-ref+rnorm(30)/2
**add more noise**
model2<-ref+rnorm(30)

1. Here the lines in bold actually have a hash (comment) at the start.  Could we include comment lines the two forms that Danny suggests?
2. It includes **rnorm**, which is quite popular, but not (yet) on the Enter keyboard.  runif is another popular one.  We don't need them all, but perhaps could include those 2?
3. I notice the gl in gl(10,3) above is also not included.  We don't need to get the keyboard "right" at this stage, but I wonder what other common data commands are also missing?
4/ I was thinking of our new-style buttons at the top, with 3 buttons, but I now wonder (with Danny's 2 options) whether there are 4, namely  **New, Wakefield, Data Frame, Data**.

**Comment by dannyparsons** [comment]
This is different because the example this is from is for using a function that doesn't have a data frame argument (taylor.diagram) so the code is structured in this way because a data frame isn't needed. To make as a dataframe it would run as a single command:

`data.frame(ref = rnorm(30, sd=2), model1 = ref+rnorm(30)/2, model2 = ref+rnorm(30))`

where we could have our standard single dialog comment.

Of course in the textbox case in the new design you can type it how you like e.g.

```
# fake some reference data
ref<-rnorm(30,sd=2)
# add a little noise
model1<-ref+rnorm(30)/2
# add more noise
model2<-ref+rnorm(30)
# put into a dataframe
dat <- data.frame(ref = ref, model1 = model1, model2 = model2)
```

This isn't good R practice because you are making intermediate variables that are not needed. There could also be confusion in functions whether you are referring to the variable ref or the column ref in dat. But we won't stop people if they want to do it this way.

**Comment by Patowhiz** [comment]
@dannyparsons @rdstern  I've not found this dialog. Assist me with the name

**Comment by rdstern** [comment]
dlgNewDataFrame.vb

**Comment by Patowhiz** [comment]
You've been calling it **File > New** instead of **File > New Data Frame** . Had me confused a bit.

**Comment by rdstern** [comment]
Sorry about that.

**Comment by Patowhiz** [comment]
@dannyparsons @rdstern The screenshot below is for when the Data option is selected. I'm still not sure where the calculator will go. The table will overlap with the script textbox/RichTextBox. I'm still contemplating the Wakefield option. What are your suggestions on this.

![sketch](https://user-images.githubusercontent.com/21179007/45480772-7270da80-b752-11e8-92ea-7fefc7dd63b1.jpg)

**Comment by dannyparsons** [comment]
I think the three main buttons at the top will be:

- Construct: the table layout you have, the default. We don't need the rows control as this is determined by the expressions for columns (we could later have checks that each expression produces something of the same length). Then rename Values to **Expression** in your table. Roger would also like a small first column in the table to show the column number. Then that looks good. The table should be quite large as expressions could be long.

- Command: This is just the big empty textbox to type an expression that evaluates to a data frame.

- Random: What you currently have a Wakefield. This needs some thought about how it will work.

Not sure if those names will stick but that's my first attempt.

My view is we don't keep anything of what is currently there. It's not something I see as particularly useful over what we are proposing, and it can be done fairly easily through the **Construct** option.

**Comment by Patowhiz** [comment]
Okay. @rdstern do you concur we remove this ?

![new data frame](https://user-images.githubusercontent.com/21179007/45482419-b1a12a80-b756-11e8-9b41-77d7435d186f.JPG)

**Comment by rdstern** [comment]
I do like the extra option of Command - which (I hope) will be very quick to do.  It might need some nifty additions later on error trapping, but halfway commands could cause issues!

I am not sure yet that I want to remove this first option with the simple blank data frame.  I have used it to create small empty data frames, roughly size of the above, and then either put data into them manually or used Prepare dialogues.  How do we get an empty data frame in the future if we get rid of this option?

I don't mind **Wakefield** becoming **Random**, though I quite liked giving credit to  the package developer.  It might be quite easy, but it is disabled for now.  I am not sure it needs to be you to do it.

**Comment by Patowhiz** [comment]
The sketch shows how the 2 interfaces will look like. Any additions? I'm still not sure what to label the table. I've labelled the big textbox as script. @dannyparsons should we use a richtextbox or just a textbox?

![sketch](https://user-images.githubusercontent.com/21179007/45540303-baf0cc80-b814-11e8-92de-1e4fc45819e0.jpg)

**Comment by rdstern** [comment]
I still wouldn't mind an extra control that is put there on all 4 options, i.e. **Number of Cases (Rows):** . This has the **up-down control**.
It is already there on the existing **File > New Data Frame** dialogue just as **Rows:**. 
I suggest it is good for the user to think of the number of rows (cases), ahead of giving the commands.  So this is compulsory, before OK is enabled.

Later (or now?) we will want to check - at least in the table format - that the different rows each produce a column of the same and right length.  This might also make that checking easier - though that's not the main reason for including this field.  

I am pretty sure @dannyparsons will disagree, so that will be interesting too!

**Comment by dannyparsons** [comment]
I can see it's useful to do that. I can't see how we can sensibly use that value without a lot of thought, other than giving an error if the row value doesn't match the number of rows being produced. But that seems annoying if what's specified gives a valid data frame. It's also not easy to calculate other than running the whole data frame command in the dialog to check the row number before its submitted, which we don't really want to do, could be quite slow.

My idea is the table command would eventually show the number of rows for each defined column, so you will already be able to see this, and an overall row number might not be needed?

If we include the row number then it must do something, open to suggestions on this, if we don't have anything concrete now I suggest leaving it out and having an open question/enhancement of how we make users think about number of rows in this dialog - I've given my suggestion for the table case above.

**Comment by Patowhiz** [comment]
@dannyparsons the will the Command option set the dataframe name and dataframe label controls invisible? I assume these will be typed by the user in the command textbox?

**Comment by dannyparsons** [comment]
No these can stay, the command box will just be the data frame e.g.

`data.frame(x = 1:10, y = 11:20)`

This is consistent with how we do commands like in the calculator and other places. It could be mentioned somewhere that the assignment isn't needed (although it will probably still work even if they did assignment)

**Comment by Patowhiz** [comment]
@rdstern @dannyparsons I'm finalising on the command option and so far I've noticed we can have a default command option that makes the Empty option redundant. Below is the default command  that I propose to be shown, on first load and on clicking reset. The output command will just be the same as   when the empty command option is used. I welcome any reservations.

![new data frame](https://user-images.githubusercontent.com/21179007/45821639-58a33a80-bcf2-11e8-88b7-27f3368f49fe.JPG)

**Comment by dannyparsons** [comment]
That's an interesting idea. I had thought `data.frame()` would be the default but I see your reasoning for this. I think I still feel that some users would find even modifying this code daunting compared to the simple nuds so I don't feel it makes Empty redundant. I do think we could make the Construct option eventually make Empty redundant though with some additions. I don't think the complexity of the design should determine what we decide to do on the dialog. 
I would probably then say that `data.frame()` should be the default command because its a better starting point for most users who will type a command and I think Empty is still needed (for now).

**Comment by Patowhiz** [comment]
@rdstern I like your idea much more than retaining your empty. We could quickly implement this as a combobox as you suggest or as a subdialog that shows "sample commands for different dataframes". The subdialog could constitute a clickable but uneditable table or listview(something similar to how our datasets are arranged in the dataset dialog) which will have columns; commands and description. We could then link the subdialog through a right click or a button just above the command textbox. When a command is double clicked then it's added to the command textbox. @dannyparsons might have reservations about this?

**Comment by Patowhiz** [comment]
In addition we could have this subdialog styled very well to make it straight forward to use and almost feel part of this dialog, like the way we made the comment subdialog.

**Comment by dannyparsons** [comment]
I don't think a whole subdialog is needed for this. Possibly an **Examples** button next to the textbox witha pull down that list the examples, when you click on one it sets thats as the text.

**Comment by Patowhiz** [comment]
I will look into the above suggestion.
@rdstern @dannyparsons this dialog is almost done. Outline all the commands and their comments that will be shown when this button is clicked.

**Comment by rdstern** [comment]
Great.  I am looking forward to seeing this.

**Comment by Patowhiz** [comment]
What about those examples? Which ones should be shown?

**Comment by rdstern** [comment]
I had my a) and b) above

Then we could have a few others that put data into a data frame, e.g. as above:

data.frame(x=1:30, y=rnorm(30,mean = 100, sd = 15), z = runif(30, min = 10, max = 30)) 

data.frame( block = gl(4,3), treat = c("C","A","B","B","C","A","A","B","C","A","C","B"), yield = (74, 68, 50,62,68,57,70,56,83, 67, 67,59).

**Comment by Patowhiz** [comment]
This is the final outlook of the form. The only thing remaining is the examples of the command option.
The default option is Command radio button.

![command](https://user-images.githubusercontent.com/21179007/45880132-a03bcc00-bdaf-11e8-82d5-a797cea24274.JPG)

![construct](https://user-images.githubusercontent.com/21179007/45880138-a762da00-bdaf-11e8-8cd6-f010951bcb99.JPG)

![empty](https://user-images.githubusercontent.com/21179007/45880155-afbb1500-bdaf-11e8-88b0-b0e4e1461abc.JPG)

**Comment by dannyparsons** [comment]
This is looking good.

1. How does the examples button work? I was expecting a drop down button on it.
2. Default should be Construct
3. On Empty, have the rows and columns options aligned vertically instead of horizontally, like it is currently. And also move rows about 20 points down and right so its further away from the edge and the top buttons.

Otherwise all looks good.

**Comment by Patowhiz** [comment]
Okay, will adjust accordingly. The examples button will show a pop-up
window when clicked. Which will have the commands and their descriptions.
I'm yet to finalize on it.

**Comment by rdstern** [comment]
Looks very good.

**Comment by dannyparsons** [comment]
I thought you could use a splitbutton for this, like we have on the toolbar.

**Comment by Patowhiz** [comment]
Okay I'll try that out.

**Comment by Patowhiz** [comment]
@dannyparsons I've just realised that the split button can only exists in a toolbar(I stand corrected if I'm wrong). In this case, I think adding a toolbar will be an overkill.

**Comment by Patowhiz** [comment]
I could kill create a nice pop up, because I still think a drop down won't be nice in terms of UI and UX. Though, I'm open to just using the drop down if it's a better alternative in your opinion

**Comment by dannyparsons** [comment]
I think you can add a toolbar of custom size and just have a split button, so it won't look like a toolbar. Have a play around but happy that you just make a decision on what you think looks best and easy to do.

**Comment by Patowhiz** [comment]
@rdstern @dannyparsons  the wakefield won't run, is there a prerquisite for this?
wakefield::r_data_theme(n = 100, data_theme = "the_works")

Then what would be comments for the other examples?

<img width="370" alt="new file" src="https://user-images.githubusercontent.com/21179007/46093135-37918c80-c1bf-11e8-92ee-4ceabb242736.png">

**Comment by dannyparsons** [comment]
What do you mean it won't run? Was there an error message? Do you have the `wakefield` package installed?

**Comment by Patowhiz** [comment]
@dannyparsons sorry, I meant, it's that specific function that @rdstern asked me to add - `wakefield::r_data_theme(n = 100, data_theme = "the_works")` . It throws an error. when you run it without explicitly importing the library in the first run(the function fails to identify function r_data_frame of the wakefield, probably it's masked). The same applies to 
`r_data_theme(, "survey")`
`r_data_theme(, "survey2")`

However a function like 
`wakefield::r_data_frame(n = 30, id, race, age, sex, hour, iq, height, died, Scoring = rnorm, Smoker = valid)` runs without explicitly importing the library on the first run.
The same occurs in RStudio.
